src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs

```diff
@@ -104,24 +103,27 @@ public virtual IApplication Start()
                         telemetrySource.WriteTelemetry("Microsoft.AspNet.Hosting.BeginRequest", new { httpContext = httpContext });
                     }
 #pragma warning restore 0618
-                    try
+                    using (logger.RequestScope(httpContext))
                     {
-                        using (logger.IsEnabled(LogLevel.Critical)
-                            ? logger.BeginScope("Request Id: {RequestId}", requestIdentifier) 
-                            : null)
+                        try
                         {
+                            logger.RequestStarting(httpContext);
                             await application(httpContext);
                         }
-                    }
-                    catch (Exception ex)
-                    {
+                        catch (Exception ex)
+                        {
 #pragma warning disable 0618
-                        if (telemetrySource.IsEnabled("Microsoft.AspNet.Hosting.UnhandledException"))
+                            if (telemetrySource.IsEnabled("Microsoft.AspNet.Hosting.UnhandledException"))
+                            {
+                                telemetrySource.WriteTelemetry("Microsoft.AspNet.Hosting.UnhandledException", new { httpContext = httpContext, exception = ex });
+                            }
+#pragma warning restore 0618
+                            throw;
+                        }
+                        finally
                         {
-                            telemetrySource.WriteTelemetry("Microsoft.AspNet.Hosting.UnhandledException", new { httpContext = httpContext, exception = ex });
+                            logger.RequestFinished(httpContext);
```

Comment:
Agreed, this is better done at the server level, but we want to get something basic done here to flesh out the experience for RC. Opened https://github.com/aspnet/Hosting/issues/443